usr/lib/common/utility.c

```diff
@@ -849,66 +849,89 @@ CK_RV get_hmac_digest(CK_ULONG mech, CK_ULONG *digest_mech, CK_BBOOL *general)
     return CKR_OK;
 }
 
-/* Compute specified SHA using either software or token implementation */
+/* Compute specified SHA or MD5 using software */
 CK_RV compute_sha(STDLL_TokData_t *tokdata, CK_BYTE *data, CK_ULONG len,
                   CK_BYTE *hash, CK_ULONG mech)
 {
-    DIGEST_CONTEXT ctx;
-    CK_ULONG hash_len;
-    CK_RV rv;
+    const EVP_MD *md;
+    unsigned int hash_len;
 
-    memset(&ctx, 0x0, sizeof(ctx));
-    ctx.mech.mechanism = mech;
+    UNUSED(tokdata);
 
-    rv = get_sha_size(mech, &hash_len);
-    if (rv != CKR_OK)
-        return rv;
+    switch (mech) {
+    case CKM_MD5:
+        hash_len = MD5_HASH_SIZE;
+        md = EVP_md5();
+        break;
+    case CKM_SHA_1:
+        hash_len = SHA1_HASH_SIZE;
+        md = EVP_sha1();
+        break;
+    case CKM_SHA224:
+    case CKM_SHA512_224:
+        hash_len = SHA224_HASH_SIZE;
+        md = EVP_sha224();
+        break;
+    case CKM_SHA256:
+    case CKM_SHA512_256:
+        hash_len = SHA256_HASH_SIZE;
+        md = EVP_sha256();
+        break;
+    case CKM_SHA384:
+        hash_len = SHA384_HASH_SIZE;
+        md = EVP_sha384();
+        break;
+    case CKM_SHA512:
+        hash_len = SHA512_HASH_SIZE;
+        md = EVP_sha512();
+        break;
+#ifdef NID_sha3_224
+    case CKM_IBM_SHA3_224:
+        hash_len = SHA3_224_HASH_SIZE;
+        md = EVP_sha3_224();
+        break;
+#endif
+#ifdef NID_sha3_256
+    case CKM_IBM_SHA3_256:
+        hash_len = SHA3_256_HASH_SIZE;
+        md = EVP_sha3_256();
+        break;
+#endif
+#ifdef NID_sha3_384
+    case CKM_IBM_SHA3_384:
+        hash_len = SHA3_384_HASH_SIZE;
+        md = EVP_sha3_384();
+        break;
+#endif
+#ifdef NID_sha3_512
+    case CKM_IBM_SHA3_512:
+        hash_len = SHA3_512_HASH_SIZE;
+        md = EVP_sha3_512();
+        break;
+#endif
+    default:
+        return CKR_MECHANISM_INVALID;
+    }
 
-    rv = sha_init(tokdata, NULL, &ctx, &ctx.mech);
-    if (rv != CKR_OK) {
-        TRACE_DEBUG("failed to create digest.\n");
-        return rv;
+    if (EVP_Digest(data, len, hash, &hash_len, md, NULL) != 1) {
```

Comment:
Did you already take a look at commit https://github.com/opencryptoki/opencryptoki/pull/416/commits/22db88ca2a086c8824326765ba738b4b5a8193ae ? This should solve such loops.

Reply:
That takes care of OpenSSL 3.0.  What about 1.1.1?  There, we just made the problem even bigger.  Since you changed common code, are now any tokens impacted that were not impacted before?

Reply:
As far as I know even before all tokens have been affected by that problem anyway. The private token encryption routines (sw_des3_cbc in sw_crypt.c) are already using the EVP interface since longer time. So this would have caused the loop already for all tokens anyway. 

Based on this, I would say the changes of this PR don't make the situation worse (but also not better for OpenSSL 1.1.1). 

I also don't know how one can force OpenSSL to NOT use a configured engine (if any). Passing NULL as engine to those functions that accept an engine doesn't help, NULL just means use whatever is configured.

Reply:
Maybe we could iterate over all registered engines via `ENGINE_get_first()`/`ENGINE_get_next()`, and remove all of them via `ENGINE_remove()` on entry to OCK code, and re-add them via `ENGINE_add()` when OCK returns back to the application? But this sounds really like a huge overhead.....